src/compiler/checker.ts

```diff
@@ -33919,7 +33919,11 @@ namespace ts {
                     const operator = operatorToken.kind;
                     if (operator === SyntaxKind.AmpersandAmpersandToken || operator === SyntaxKind.BarBarToken || operator === SyntaxKind.QuestionQuestionToken) {
                         if (operator === SyntaxKind.AmpersandAmpersandToken) {
-                            const parent = walkUpParenthesizedExpressions(node.parent);
+                            let parent = node.parent;
+                            while (parent.kind === SyntaxKind.ParenthesizedExpression
+                                || isBinaryExpression(parent) && (parent.operatorToken.kind === SyntaxKind.AmpersandAmpersandToken || parent.operatorToken.kind === SyntaxKind.BarBarToken)) {
+                                parent = parent.parent;
+                            }
                             checkTestingKnownTruthyCallableOrAwaitableType(node.left, isIfStatement(parent) ? parent.thenStatement : undefined);

```

Comment:
> Why do we allow using an always truthy expression in a condition if it is later used in the if's body?

This is the heuristic we’ve always employed to avoid erroring on defensive checks:

```ts
declare function doSomething(): boolean;

if (doSomething) { // Type indicates should be present, but clearly checked on purpose
  doSomething(); 
}

if (doSomething) { // Probably a mistake
  return;
}
```

> Why are we now only skipping over parenthesis and && and || expressions to find the parent?

So, the feature is trying to detect places where the user checked the truthiness of a function in a conditional rather than the result of calling that function. So the relevant question is, if we have an AST like:

```
IfStatement
  condition:
    ???
      BinaryExpression (&&)
  then:
    Block
      ...
```

what are the possible structures of `???` such that we’d want to check the left and right of this `&&` expression for uncalled functions? Previously, the answer was just “parenthesized expressions,” which presents a problem for this:

```ts
if (f1 && f2 && f3) {
  f1();
  f2();
  f3();
}
```

When we’re on the first `&&` token, the `parent` is the higher `&&` binary expression, which means we run the uncalled function check for `f1` without sending the `if` body along to see if it’s used in there. In general, in a chain of `&&`s, only the last two ever got correctly checked since the last `&&` token belongs to the highest-level binary expression such that its `parent` was the IfStatement.

On the other hand, the answer to what `???` can be clearly isn’t “everything” either, because you need to rule out stuff like

```ts
if (f1 !== doSomethingWith(f2 && f3)) {
  f2();
  f3();
}
```

Here, `f2` and `f3` are nested in a condition, but their truthiness is not actually being tested by the `if`, so it’s not relevant to this check.

It could be that there are more constructs we should care about and just `&&`, `||`, and parens is an incomplete list, but it’s more complete than before, and I was able to convince myself that any arbitrary chaining of those three kinds of expressions directly inside a condition is necessarily concerned with the truthiness of all the leaves.